src/params_loader.rs

```diff
@@ -15,9 +17,82 @@ pub trait PipesParamsLoader {
     /// to create a PipesContext or should instead return a mock.
     fn is_dagster_pipes_process(&self) -> bool;
     /// Load params passed by the orchestration-side context injector.
-    fn load_context_params(&self) -> Map<String, Value>;
+    fn load_context_params(&self) -> Result<Map<String, Value>, ParamsError>;
     /// Load params passed by the orchestration-side message reader.
-    fn load_message_params(&self) -> Map<String, Value>;
+    fn load_message_params(&self) -> Result<Map<String, Value>, ParamsError>;
+}
+
+#[derive(Debug, Error)]
+#[error("{} parameter \"{}\" is {}", .origin, .param, .source)]
+pub struct ParamsError {
```

Comment:
I'd normally migrate this error to a separate module as I typically work with small to medium-sized codebases, but I'm leaving it here unless you think otherwise.